pool/opaque.go

```diff
@@ -0,0 +1,106 @@
+package pool
+
+import (
+	"fmt"
+	"reflect"
+	"unsafe"
+)
+
+// OpaquePoolOptions specify options for OpaquePool.
+type OpaquePoolOptions struct {
+	Construct func(ptr interface{})
+	Size      int
+	Type      reflect.Type
+}
+
+// OpaquePool represents an object pool which is opaque to the Go GC.
+type OpaquePool struct {
+	heap       []uint8
+	free       chan uintptr
+	opts       OpaquePoolOptions
+	step, size uintptr
+}
+
+// NewOpaquePool constructs a new OpaquePool.
+func NewOpaquePool(opts OpaquePoolOptions) *OpaquePool {
+	p := &OpaquePool{
+		free: make(chan uintptr, opts.Size),
+		opts: opts,
+		// TODO(@kobolog): alignment & padding.
+		step: hsz + opts.Type.Size(),
+	}
+
+	p.size = uintptr(p.opts.Size * int(p.step))
+
+	// Heap is a slice of uint8 large enough to fit opts.Size objects
+	// of type struct { hdr; T }.
+	p.heap = make([]byte, p.size)
+
+	if p.opts.Construct == nil {
+		p.opts.Construct = func(interface{}) {}
+	}
+
+	p.init()
+
+	return p
+}
+
+func (p *OpaquePool) String() string {
+	return fmt.Sprintf(
+		"OpaquePool[%d/%d: %s, %p-%p, %d]", len(p.free), cap(p.free),
+		p.opts.Type, &p.heap[0], &p.heap[p.size-1], p.size)
+}
+
+type hdr struct {
+	// Offset from the beginning of the heap to the object.
+	offset uintptr
+}
+
+const (
+	hsz = unsafe.Sizeof(*(*hdr)(nil))
+)
+
+func (p *OpaquePool) init() {
+	for i := uintptr(0); i < p.size; i += p.step {
+		hdr := (*hdr)(unsafe.Pointer(&p.heap[i]))
+		hdr.offset = i + hsz
+		ptr := unsafe.Pointer(&p.heap[hdr.offset])
+
+		p.opts.Construct(reflect.NewAt(p.opts.Type, ptr).Interface())
+
+		// Alternatively, we can use a channel of interfaces, but it
+		// would keep opts.Size objects rooted in the channel.
+		// We aim to avoid allocating unnecessary GC-visible objects.
+		p.free <- hdr.offset
+	}
+}
+
+// Get provides an object from the pool.
+func (p *OpaquePool) Get() interface{} {
```

Comment:
Seems we don't reset the object when getting it from the pool or returning it to the pool. Would it be better to add a comment saying the callers should reset the state of the object returned from pool?